Yosemite/Yosemite/Tools/POS/POSOrderService.swift

```diff
@@ -5,15 +5,79 @@ import class WooFoundation.CurrencyFormatter
     }
 }
 
+extension [POSCartItem] {
+    public func matches(order: Order?) -> Bool {
+        guard let order else {
+            return self.isEmpty
+        }
+
+        let consolidatedCartItems = self.reduce(into: [POSCartItem]()) { partialResult, nextItem in
+            if let matchingIndex = partialResult.firstIndex(where: { $0.item.isEqual(to: nextItem.item) }) {
+                let itemToUpdate = partialResult[matchingIndex]
+                partialResult[matchingIndex] = POSCartItem(item: itemToUpdate.item, quantity: itemToUpdate.quantity + nextItem.quantity)
+            } else {
+                partialResult.append(nextItem)
+            }
+        }
+
+        let consolidatedOrderItems = order.items.reduce(into: [OrderItem]()) { partialResult, nextItem in
+            if let matchingIndex = partialResult.firstIndex(where: { $0.productID == nextItem.productID && $0.variationID == nextItem.variationID }) {
+                let itemToUpdate = partialResult[matchingIndex]
+                partialResult[matchingIndex] = itemToUpdate.copy(quantity: itemToUpdate.quantity + nextItem.quantity)
+            } else {
+                partialResult.append(nextItem)
+            }
+        }
+
+        guard consolidatedCartItems.count == consolidatedOrderItems.count else {
+            return false
+        }
+
+        for cartItem in consolidatedCartItems {
+            guard consolidatedOrderItems.contains(where: { orderItem in
+                cartItem.item.matches(orderItem: orderItem) && cartItem.quantity == orderItem.quantity
+            }) else {
+                return false
+            }
+        }
+        return true
+    }
+
+    func createGroupedOrderSyncProductInputs() -> [OrderSyncProductInput.ProductType: OrderSyncProductInput] {
+        let orderSyncProductInputs = self.map { $0.item.toOrderSyncProductInput(quantity: $0.quantity) }
+
+        // Group items by their `product`, which is actually a `ProductType` enum, representing a product or variation,
+        // with an associated value for the underlying item.
+        let groupedItems = Dictionary(grouping: orderSyncProductInputs, by: { $0.product })
+
+        // Convert each group into a single `OrderSyncProductInput`
+        let output = groupedItems.compactMapValues { items -> OrderSyncProductInput? in
+            guard let firstItem = items.first else { return nil }
+
+            // Aggregate the quantity for this item
+            let totalQuantity = items.reduce(Decimal(0)) { $0 + $1.quantity }
+
+            // Return a copy of the first item, with the aggregate quantity
+            return OrderSyncProductInput(product: firstItem.product,
+                                         quantity: totalQuantity,
+                                         discount: firstItem.discount,
+                                         baseSubtotal: firstItem.baseSubtotal,
+                                         bundleConfiguration: firstItem.bundleConfiguration)
```

Comment:
We don't really pass them either; `toOrderSyncProductInput` does the work of creating the `OrderSyncProductInput` that we're copying from here, and only provides product and quantity. Since we're making a copy though, it seems safer to use the full init here, so we don't miss something off the copy in future.

I left off the ID as it doesn't make sense to say the new object is the same as the first object passed in.

There is some risk that this is incorrect; baseSubtotal looks fine as it's a stand-in for price, so it's not related to the quantity. Discount, I'm less sure about. Perhaps this should strictly be aggregate by product and discount... but let's tackle that if/when we allow item discounts.